### PR TITLE
support hiredis >= 1.0

### PR DESCRIPTION
**Pull request by 5c30**
i'm on a system with hiredis 1.0.2 installed. i ran into problems with Protocol::Redis::XS (segfaults and "bizarre copy of UNKNOWN" errors from perl), so i hacked some stuff together to get it working with the newer version.

i have pretty much no idea what i'm doing in XS, so please review carefully. this change "works for me," but it's likely i've done horrible things i'm unaware of. i added #if guards around the changes to hopefully keep it working on older versions of hiredis, but haven't been able to test that.

hope it's helpful.

### CODE REVIEW

**Comment by Grinnz** [comment]
I also cannot evaluate the XS very well but will test the changes with different versions of hiredis. Thank you for the contribution.

**Comment by eleksir** [comment]
me-too :)
I've got
`hiredis-1.0.0-r0 x86_64 {hiredis} (BSD-3-Clause) [installed]`
on alpine linux 3.12 x86_64. During installation with cpanm I face almost same issue - module build fails at testing stage with:
```
Configuring Protocol-Redis-XS-0.07                                                                                                                                           
Running Makefile.PL                                                                                                                                                          
Checking if your kit is complete...                                                                                                                                          
Looks good                                                                                                                                                                   
Generating a Unix-style Makefile                                                                                                                                             
Writing Makefile for Protocol::Redis::XS                                                                                                                                     
Writing MYMETA.yml and MYMETA.json                                                                                                                                           
-> OK                                                                                                                                                                        
Checking dependencies from MYMETA.json ...                                                                                                                                   
Checking if you have XSLoader 0 ... Yes (0.30)                                                                                                                               
Checking if you have parent 0 ... Yes (0.238)                                                                                                                                
Checking if you have Test::More 0.98 ... Yes (1.302188)                                                                                                                      
Checking if you have XS::Object::Magic 0 ... Yes (0.05)                                                                                                                      
Checking if you have ExtUtils::MakeMaker 7.44 ... Yes (7.44)                                                                                                                 
Checking if you have Protocol::Redis 1.0011 ... Yes (1.0011)                                                                                                                 
Checking if you have Protocol::Redis::Test 0 ... Yes (undef)                                                                                                                 
Building and testing Protocol-Redis-XS-0.07                                                                                                                                  
cp lib/Protocol/Redis/XS.pm blib/lib/Protocol/Redis/XS.pm                                                                                                                    
Running Mkbootstrap for XS ()                                                                                                                                                
chmod 644 "XS.bs"                                                                                                                                                            
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Protocol/Redis/XS/XS.bs 644                                                                 
"/usr/bin/perl" "/usr/share/perl5/core_perl/ExtUtils/xsubpp"  -typemap '/usr/share/perl5/core_perl/ExtUtils/typemap' -typemap '/var/lib/aleesa-misc.new/vendor_perl/lib/perl5
mv XS.xsc XS.c                                                                                                                                                               
cc -c  -I/var/lib/aleesa-misc.new/vendor_perl/lib/perl5/x86_64-linux-thread-multi/Alien/hiredis/Install -I/usr/include/hiredis -D_FILE_OFFSET_BITS=64 -I/var/lib/aleesa-misc.
XS.xs:114:3: warning: initialization of 'void * (*)(const redisReadTask *, size_t)' {aka 'void * (*)(const redisReadTask *, long unsigned int)'} from incompatible pointer ty
  114 |   createArrayObjectSV,                                                                                                                                               
      |   ^~~~~~~~~~~~~~~~~~~                                                                                                                                                
XS.xs:114:3: note: (near initialization for 'perlRedisFunctions.createArray')                                                                                                
XS.xs:116:3: warning: initialization of 'void * (*)(const redisReadTask *, double,  char *, size_t)' {aka 'void * (*)(const redisReadTask *, double,  char *, long unsigned i
  116 |   createNilObjectSV,                                                                                                                                                 
      |   ^~~~~~~~~~~~~~~~~                                                                                                                                                  
XS.xs:116:3: note: (near initialization for 'perlRedisFunctions.createDouble')                                                                                               
XS.xs:117:3: warning: initialization of 'void * (*)(const redisReadTask *)' from incompatible pointer type 'void (*)(void *)' [-Wincompatible-pointer-types]                 
  117 |   freeReplyObjectSV                                                                                                                                                  
      |   ^~~~~~~~~~~~~~~~~                                                                                                                                                  
XS.xs:117:3: note: (near initialization for 'perlRedisFunctions.createNil')                                                                                                  
rm -f blib/arch/auto/Protocol/Redis/XS/XS.so                                                                                                                                 
LD_RUN_PATH="/usr/lib" cc  -shared -Os -fomit-frame-pointer -L/usr/local/lib -fstack-protector-strong  XS.o  -o blib/arch/auto/Protocol/Redis/XS/XS.so  \                    
   -lhiredis   \
                                                                                                                                                                             
chmod 755 blib/arch/auto/Protocol/Redis/XS/XS.so                                                                                                                             
Manifying 1 pod document                                                                                                                                                     
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Protocol/Redis/XS/XS.bs 644                                                                 
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t          
    # Looks like you planned 43 tests but ran 13.                                                                                                                            
                                                                                                                                                                             
#   Failed test 'Protocol::Redis APIv1 ok'                                                                                                                                   
#   at /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/Protocol/Redis/Test.pm line 51.                                                                                        
Bizarre copy of UNKNOWN in scalar assignment at /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/Protocol/Redis/Test.pm line 119.                                              
# Looks like your test exited with 255 just after 2.                                                                                                                         
t/redis.t ..                                                                                                                                                                 
Dubious, test returned 255 (wstat 65280, 0xff00)                                                                                                                             
Failed 1/2 subtests                                                                                                                                                          
t/xs.t ..... ok                                                                                                                                                              
                                                                                                                                                                             
Test Summary Report                                                                                                                                                          
-------------------                                                                                                                                                          
t/redis.t (Wstat: 65280 Tests: 2 Failed: 1)                                                                                                                                  
  Failed test:  2                                                                                                                                                            
  Non-zero exit status: 255                                                                                                                                                  
Files=2, Tests=8,  0 wallclock secs ( 0.04 usr  0.01 sys +  0.19 cusr  0.03 csys =  0.27 CPU)                                                                                
Result: FAIL                                                                                                                                                                 
Failed 1/2 test programs. 1/8 subtests failed.                                                                                                                               
make: *** [Makefile:1080: test_dynamic] Error 255                                                                                                                            
-> FAIL Testing Protocol-Redis-XS-0.07 failed but installing it anyway.                                                                                                      
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Protocol/Redis/XS/XS.bs 644                                                                 
Manifying 1 pod document                                                                                                                                                     
Files found in blib/arch: installing files in blib/lib into architecture dependent library tree                                                                              
Installing /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/x86_64-linux-thread-multi/auto/Protocol/Redis/XS/XS.so                                                             
Installing /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/x86_64-linux-thread-multi/Protocol/Redis/XS.pm                                                                     
Installing /var/lib/aleesa-misc.new/vendor_perl/man/man3/Protocol::Redis::XS.3pm                                                                                             
Appending installation info to /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/x86_64-linux-thread-multi/perllocal.pod                                                        
-> OK                                                                                                                                                                        
Successfully installed Protocol-Redis-XS-0.07                                                                                                                                
Installing /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/x86_64-linux-thread-multi/.meta/Protocol-Redis-XS-0.07/install.json                                                
Installing /var/lib/aleesa-misc.new/vendor_perl/lib/perl5/x86_64-linux-thread-multi/.meta/Protocol-Redis-XS-0.07/MYMETA.json
```

In "production" i did not encounter any segfaults, but it looks like because I'm using redis as pub-sub only.

Anyway i think that tests must complete successfully, that is their purpose, right?

File: XS.xs [hunk 1]
```diff
@@ -98,6 +107,20 @@ static void *createIntegerObjectSV(const redisReadTask *task, long long value)
...
 }
 
+#if HIREDIS_MAJOR > 0
+static void *createDoubleObjectSV(const redisReadTask *task, double value,
+  char* str, size_t len)
+{
+  dTHXREDIS(task);
+
+  SV *sv = newSVpvn(str, len);
+  sv_setnv(sv, value);
```
Comment (Grinnz) [inline XS.xs:117]: Suggested by @mauke to preserve the string value, as sv_setnv clears it:
```suggestion
  SvUPGRADE(sv, SVt_PVNV);
  SvNV_set(sv, value);
  SvNOK_on(sv);
```

**Comment by Grinnz** [comment]
Thank you! This should allow building against hiredis 1.0+, full support for RESP3 should come later and under a different API specification. #12

**Pull request merged**